source/agora/consensus/ValidatorSet.d

```diff
@@ -98,27 +102,43 @@ public class ValidatorSet
             return false;
         }
 
-        // check if already exists
-        if (this.hasEnrollment(enroll.utxo_key))
-        {
-            log.info("Rejected already existing validator, Data was: {}",
-                enroll);
-            return false;
-        }
-
         try
         {
+            const bool exist = this.hasEnrollment(enroll.utxo_key);
+            if (exist)
+            {
+                // this checks that the re-enrollment is requested past the half
+                // of the validator cycle.
+                if (block_height - this.getEnrolledHeight(enroll.utxo_key) <
```

Comment:
Shouldn't this check be in the validation part, not here?

Reply:
Sorry for the late answer.
I think the validation part is for validating the enrollment itself. The checking based on the relation with other objects/values like `block_height` or  `enroll_height` can be here because the enrollment has no height information and the validation part is also used in adding the enrollment in the enrollment pool.